### PR TITLE
feat: implement utility token

### PR DESCRIPTION
**Pull request by egonspace**
- revise GameToken (refer 9cadfafb437ad648318790a7836bea8df7842ce9)
- add UtilityToken
  - unlimited supply
  - minter role
- add ItemNFT (from SampleNFT of iskra-contracts)
- revise directory tree (created token/ERC20, token/ERC721, token/ERC1155)
- upgrade OZ version, solidity compiler version

### CODE REVIEW

File: contracts/token/ERC20/UtilityToken.sol [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+// @author iskra.world dev team
+
+pragma solidity ^0.8.2;
+
+import "@openzeppelin/contracts/access/Ownable2Step.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import "../../kip/IKIP13.sol";
```
Comment (jyeoniii) [inline contracts/token/ERC20/UtilityToken.sol]: how about using `EIP` instead of `KIP`?

File: doc/kor/game_token.md [hunk 2]
```diff
@@ -1,17 +1,17 @@
 # ISKRA Game Token(ERC20) Contract
 
 ## 개요
-- `contracts/token/GameToken.sol` 파일은 ERC20 표준을 따르는 샘플 토큰 구현체로서, 이름(name), 심볼(symbol), 총 통화량(initial supply)을 배포 때 정할 수 있도록 구현된 토큰입니다.
+- `contracts/token/ERC20/GameToken.sol` 파일은 ERC20 표준을 따르는 샘플 토큰 구현체로서, 이름(name), 심볼(symbol), 총 통화량(initial supply)을 배포 때 정할 수 있도록 구현된 토큰입니다.
 - 배포시 초기 발행량이 정해지면, 추가 mint나 burn 기능이 불가능한 고정 통화량 토큰입니다.
 - decimal은 18로 고정입니다.
 - 초기 배포되면 owner가 총 통화량을 모두 소유하게 됩니다.
 - IERC20, IKIP13, IKIP7, IERC165 표준 인터페이스를 따릅니다.
```
Comment (brew0722) [inline doc/kor/game_token.md]: still remains here about IKIP13, IKIP7

File: doc/kor/utility_token.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
+# ISKRA Utility Token(ERC20) Contract
+
+## 개요
+- `contracts/token/ERC20/UtilityToken.sol` 파일은 ERC20 표준을 따르는 샘플 토큰 구현체로서, 이름(name), 심볼(symbol), 최초 minter를 배포 때 정할 수 있도록 구현된 토큰입니다.
+- 배포시 초기 발행량은 0이며, minter 권한이 있는 account로 mint나 burn을 할 수 있습니다.
+- minter는 무제한 발행이 가능합니다.
+- owner는 minter를 추가/삭제할 수 있습니다.
+- decimal은 18로 고정입니다.
+- 초기 배포되면 owner가 총 통화량을 모두 소유하게 됩니다.
+- IERC20, IKIP13, IKIP7, IERC165 표준 인터페이스를 따릅니다.
```
Comment (brew0722) [inline doc/kor/utility_token.md]: ditto

File: doc/kor/utility_token.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
+# ISKRA Utility Token(ERC20) Contract
+
+## 개요
+- `contracts/token/ERC20/UtilityToken.sol` 파일은 ERC20 표준을 따르는 샘플 토큰 구현체로서, 이름(name), 심볼(symbol), 최초 minter를 배포 때 정할 수 있도록 구현된 토큰입니다.
+- 배포시 초기 발행량은 0이며, minter 권한이 있는 account로 mint나 burn을 할 수 있습니다.
+- minter는 무제한 발행이 가능합니다.
+- owner는 minter를 추가/삭제할 수 있습니다.
+- decimal은 18로 고정입니다.
+- 초기 배포되면 owner가 총 통화량을 모두 소유하게 됩니다.
```
Comment (brew0722) [inline doc/kor/utility_token.md]: Based on the description above, a UtilityToken must have an initialSupply of 0.
However, here it says that there is a supply owned by owner.

**Review by brew0722 — APPROVED** [review]
LGTM

File: contracts/token/ERC20/GameToken.sol [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    function approveAndCall(address, uint256)
+        external
+        pure
+        override
+        returns (bool)
+    {
+        require(
+            false,
+            "GameToken: does not support approveAndCall due to security issue of approve"
+        );
```
Comment (whylee259) [inline contracts/token/ERC20/GameToken.sol]: nit;
```suggestion
        revert(
            "GameToken: does not support approveAndCall due to security issue of approve"
        );
```

File: contracts/token/ERC20/GameToken.sol [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        uint256 amount
+    ) internal override {
+        require(
+            to != address(this),
+            "GameToken: cannot transfer tokens to the token contract"
+        );
+        super._beforeTokenTransfer(from, to, amount);
+    }
+
+    function _checkOnTransferReceived(
```
Comment (whylee259) [inline contracts/token/ERC20/GameToken.sol:122]: This is a newer version of `_checkOnTransferReceived`.
> https://github.com/vittominacori/erc1363-payable-token/blob/master/contracts/token/ERC1363/ERC1363.sol#L117-L134

This adds handling for error messages.

File: contracts/token/ERC721/ItemNFT.sol [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+// @author iskra.world dev team
+
+pragma solidity ^0.8.2;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+// ItemNFT features
+//  - ERC721 basic functions
+contract ItemNFT is ERC721, Ownable2Step {
```
Comment (whylee259) [inline contracts/token/ERC721/ItemNFT.sol:12]: discussion: burnning
For the case of Mystery Box's reveal, it needs to be determined how should be done for the collected NFT.
- Add a burn function and grant permission to the reveal contract
- Send them to the account like 0xdead.
- Simply keep them

Comment (egonspace) [reply contracts/token/ERC721/ItemNFT.sol:12]: I think we should apply MultiToken's burn policy equally to ItemNFT. However, a separate PR is needed.

File: contracts/token/ERC721/ItemNFT.sol [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    string internal baseURI = "";
```
Comment (whylee259) [inline contracts/token/ERC721/ItemNFT.sol]:
```suggestion
    string internal _baseURI = "";
```

File: contracts/token/ERC20/GameToken.sol [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+//  - not mintable, not burnable
+//  - IERC1363Receiver feature
+//  - defense for transferring tokens to the token contract itself
+contract GameToken is IERC20Metadata, ERC20, IERC1363, Ownable2Step {
+    using Address for address;
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        uint256 _initialSupply
```
Comment (whylee259) [inline contracts/token/ERC20/GameToken.sol]:
```suggestion
        string memory name_,
        string memory symbol_,
        uint256 initialSupply
```

File: contracts/token/ERC20/UtilityToken.sol [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    using Address for address;
+
+    mapping(address => bool) public minters;
+
+    event MinterAdded(address indexed newMinter);
+    event MinterRemoved(address indexed minter);
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
```
Comment (whylee259) [inline contracts/token/ERC20/UtilityToken.sol]:
```suggestion
        string memory name_,
        string memory symbol_,
```

**Review by whylee259 — APPROVED** [review]
Adding some minor comments. Overall, LGTM

File: contracts/token/ERC721/ItemNFT.sol (hunk 8 above)
Comment (egonspace) [reply contracts/token/ERC721/ItemNFT.sol]: `_baseURI` is already used for the `_baseURI()` internal function, so I renamed it to `mBaseURI`.

**Pull request merged**